File: Server/models/Budget.js
import mongoose from 'mongoose';

const budgetSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  name: {
    type: String,
    required: true,
    trim: true,
  },
  category: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Category',
    required: true,
  },
  // Budget period
  period: {
    type: String,
    enum: ['monthly', 'weekly', 'yearly', 'custom'],
    default: 'monthly',
  },
  // For custom period
  startDate: {
    type: Date,
    default: null,
  },
  endDate: {
    type: Date,
    default: null,
  },
  // Budget limit
  limit: {
    type: Number,
    required: true,
    min: 0.01,
  },
  // Alert threshold as a percentage (e.g. 80 means alert at 80% spent)
  alertThreshold: {
    type: Number,
    default: 80,
    min: 1,
    max: 100,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  // rollover: carry unused budget to next month
  rollover: {
    type: Boolean,
    default: false,
  },
}, {
  timestamps: true,
});

budgetSchema.index({ user: 1, isActive: 1 });
budgetSchema.index({ user: 1, category: 1, isActive: 1 });

const Budget = mongoose.model('Budget', budgetSchema);
export default Budget;

File: Android/app/src/main/java/com/finora/app/data/model/Budget.kt
package com.finora.app.data.model

import com.google.gson.annotations.SerializedName

data class Budget(
    @SerializedName("_id") val id: String = "",
    val user: String = "",
    val name: String = "",
    val category: Any? = null, // Can be String ID or populated Category object
    val period: String = "monthly", // monthly, weekly, yearly, custom
    val startDate: String? = null,
    val endDate: String? = null,
    val limit: Double = 0.0,
    val alertThreshold: Int = 80,
    val isActive: Boolean = true,
    val rollover: Boolean = false,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val categoryId: String?
        get() = when (category) {
            is String -> category
            is Map<*, *> -> category["_id"] as? String
            else -> null
        }

    val categoryName: String
        get() = when (category) {
            is Map<*, *> -> (category["name"] as? String) ?: name
            else -> name
        }

    val categoryColor: String
        get() = when (category) {
            is Map<*, *> -> (category["color"] as? String) ?: "#4F46E5"
            else -> "#4F46E5"
        }
}

// Server Response for GET /api/budgets/with-spend
data class BudgetWithSpend(
    @SerializedName("_id") val id: String = "",
    val name: String = "",
    val category: Any? = null,
    val period: String = "monthly",
    val limit: Double = 0.0,
    val alertThreshold: Int = 80,
    val spent: Double = 0.0,
    val remaining: Double = 0.0,
    val percentage: Double = 0.0,
    val status: String = "Safe" // Safe, Warning, Exceeded
) {
    val categoryName: String
        get() = when (category) {
            is Map<*, *> -> (category["name"] as? String) ?: name
            else -> name
        }

    val categoryColor: String
        get() = when (category) {
            is Map<*, *> -> (category["color"] as? String) ?: "#4F46E5"
            else -> "#4F46E5"
        }
}
